modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java

```diff
@@ -0,0 +1,261 @@
+
+    // Make sure either one of tags or flavors are provided
+    if (StringUtils.isBlank(sourceTagsOption) && StringUtils.isBlank(sourceFlavorsOption)) {
+      logger.info("No source tags or flavors have been specified, not matching anything");
+      return createResult(mediaPackage, Action.CONTINUE);
+    }
+
+    List<String> targetFlavors = asList(targetFlavorsOption);
+    String[] targetTags = StringUtils.split(targetTagsOption, sectionSeparator);
+    AbstractMediaPackageElementSelector<Track> elementSelector = new TrackSelector();
+
+    // Select the source flavors
+    for (String flavor : asList(sourceFlavorsOption)) {
+      try {
+        elementSelector.addFlavor(MediaPackageElementFlavor.parseFlavor(flavor));
+      } catch (IllegalArgumentException e) {
+        throw new WorkflowOperationException(String.format("Source flavor '%s' is malformed", flavor));
+      }
+    }
+
+    // Select the source tags
+    for (String tag : asList(sourceTagsOption)) {
+      elementSelector.addTag(tag);
+    }
+
+    // Find the encoding profile - should only be one
+    EncodingProfile profile = composerService.getProfile(encodingProfile);
+    if (profile == null) {
+      throw new WorkflowOperationException(String.format("Encoding profile '%s' was not found", encodingProfile));
+    }
+    // Look for elements matching the tag
+    Collection<Track> sourceTracks = elementSelector.select(mediaPackage, false);
+    if (sourceTracks.isEmpty()) {
+      logger.info("No matching tracks found");
+      return createResult(mediaPackage, Action.CONTINUE);
+    }
+
+    long totalTimeInQueue = 0;
+    Map<Job, Track> encodingJobs = new HashMap<>();
+    for (Track track : sourceTracks) {
+      logger.info("Encoding track {} using encoding profile '{}'", track, profile);
+      // Start encoding and wait for the result
+      encodingJobs.put(composerService.demux(track, profile.getIdentifier()), track);
+    }
+
+    // Wait for the jobs to return
+    if (!waitForStatus(encodingJobs.keySet().toArray(new Job[encodingJobs.size()])).isSuccess()) {
+      throw new WorkflowOperationException("One of the encoding jobs did not complete successfully");
+    }
+
+    // Process the result
+    for (Map.Entry<Job, Track> entry : encodingJobs.entrySet()) {
+      Job job = entry.getKey();
+      Track sourceTrack = entry.getValue();
+
+      // add this receipt's queue time to the total
+      totalTimeInQueue += job.getQueueTime();
+
+      // it is allowed for compose jobs to return an empty payload. See the EncodeEngine interface
+      if (job.getPayload().length() <= 0) {
+        logger.warn("No output from Demux operation");
+        continue;
+      }
+
+      List<Track> composedTracks = (List<Track>) MediaPackageElementParser.getArrayFromXml(job.getPayload());
+      if (composedTracks.size() != targetFlavors.size() && targetFlavors.size() != 1) {
```

Comment:
Shouldn't we be checking if targetFlavors.size() is not 1 prior to dispatching the jobs?

Reply:
No, that's ok. if it is 1, the same target flavor will be used for all outputs